packages/server/src/adapters/fetch/content-type/json/index.ts

```diff
@@ -16,79 +17,75 @@ export interface FetchHTTPContentTypeHandler<TRouter extends AnyRouter>
 
 export const getFetchHTTPJSONContentTypeHandler: <
   TRouter extends AnyRouter,
->() => FetchHTTPContentTypeHandler<TRouter> = () => ({
-  name: 'fetch-json',
-  isMatch: (headers) => {
-    return !!headers.get('content-type')?.startsWith('application/json');
-  },
-  getInputs: async (opts, info) => {
-    async function getRawProcedureInputOrThrow() {
-      const { req } = opts;
+>() => FetchHTTPContentTypeHandler<TRouter> = () => {
+  const rawInputCache = createConcurrentCache();
```

Comment:
The bug was that when we map over a cache, all the promises which start waiting on this body start at once and pass control to the next one while waiting on the body. So yes we were consuming this N times at once.

By using a concurrent cache the first iteration will synchronously cache the promise and then future batches will read the current promise, additionally we can move the batch caching inside here (see next line) which is better for simplicity in the adapter handlers